sam/exploration/signalaligner.py

```diff
@@ -0,0 +1,259 @@
+import numpy as np
+import pandas as pd
+
+
+class SignalAligner:
+    """Allows temporal alignment of two signals or dataframes containing two signals
+    to be aligned. We assume both signals have the same sampling frequency. For now
+    there is no timestamp based alignment, we simply use the cross-correlation of the
+    to be aligned signals (thereby assuming equal sampling frequencies).
+
+    Parameters
+    ----------
+    signal_one : np.ndarray (default=None)
+    signal_two : np.ndarray (default=None)
+
+    Examples
+    --------
+    # Example 1
+    >>> import numpy as np
+    >>> from sam.exploration.signalaligner import SignalAligner
+
+    >>> signal_one = np.array([0, 1, 2, 3, 4, 3, 2, 1, 0])
+    >>> signal_two = np.array([0, 0, 0, 0, 0, 1, 2, 3, 4, 3, 2])
+    >>> offset, _ = SignalAligner.align_signals(signal_one, signal_two)
+
+    >>> print('Offset =', offset)
+
+    # Example 2
+    >>> N1 = 20
+    >>> N2 = 30
+    >>> N_aligned = 15
+
+    >>> lat = np.random.randn(N_aligned) + np.random.standard_normal(N_aligned)
+
+    >>> lat1 = np.random.randn(N1)
+    >>> i1 = np.random.randint(N1 - N_aligned)
+    >>> lat1[i1: i1 + N_aligned] = lat
+
+    >>> lat2 = np.random.randn(N2)
+    >>> i2 = np.random.randint(N2 - N_aligned)
+    >>> lat2[i2: i2 + N_aligned] = lat
+
+    >>> df1 = pd.DataFrame({'data': np.random.randn(N1), 'lat': lat1})
+    >>> df2 = pd.DataFrame({'data': np.random.randn(N2), 'lat': lat2})
+
+    >>> sa = SignalAligner()
+    >>> df, offset = sa.align_dataframes(df1, df2, col1, col2, reference=0)
+    """
+
+    def __init__(self, signal_one=None, signal_two=None):
+
+        self.signal_one = signal_one
+        self.signal_two = signal_two
+
+        if (signal_one is not None) and (signal_two is not None):
+            signal_one, signal_two = self._preprocess_signals(signal_one, signal_two)
```

Comment:
I wouldn't do preprocessing in the init, I think it's better to create a fit function or maybe even better: base it on the sklearn transformer interface

Reply:
mhh, I like the idea of basing it on the sklearn transformer interface. But it is a little odd to use it here, if usually you try to align two different signals or two different dataframes. It would be more natural to have a single dataframe as an input to the sklearn based transformer and perform some alignment on its columns. But that is a strange use case, I think it's more common to have two different signals (with different signal lengths) or two different dataframes (with different lengths).

We could pass two signals or dataframes as inputs to the fit method I suppose, but that is somewhat unexpected from a sklearn transformer. For now I would opt for keeping the general current structure, adding a fit method for preprocessing, what do you think?